### ISSUE TITLE
Erroneous "can't auth publish" with vmq_diversity MySQL DB

### ISSUE THREAD

**Issue by DouglasWWolf** [vernemq/vernemq#1164]
Reproducible issue regarding faulty publish authorization:

VMQ Version: 1.7.1
O/S:  Centos 7
Auth database: MySQL
Cluster size: 1 (i.e., standalone)

There is an authorization bug that appears when load-testing.

I have 8000 unique users in my MySQL auth table, all of them programatically generated where the client-id, username, and password are all of the form "sim10001", "sim10002", etc.  the "publish_acl" for each of those clients is the same, and looks like this:

[{'pattern':'node_ts/+/%c'}]

I have python code that connects all 8000 users to VerneMQ over the course of 30 or 40 seconds.   All connections receive a CONNACK, and work as expected.

Over the next 10 minutes, each of those 8000 connected clients sends 10 message spaced roughly 60 seconds apart, on topic "node_ts/x/<client_id>" where client_id is "sim10001", "sim10002", etc.

When I run the process described above, VerneMQ will pick one or two users seemingly at random, and decide that their attempts to publish are not authorized.   I see these messages in the log:

"2019-05-05 23:28:28.376 [error] <0.1570.2>@vmq_mqtt_fsm:auth_on_publish:718 can't auth publish [<<"sim14442">>,{[],<<"sim14442">>},0,[<<"node_ts">>,<<"2">>,<<"sim14442">>],<<"hello">>,false] due to not_authorized"

As I mentioned, the "unlucky" user (or occasionally two users, or occasionally none) are perfectly valid users, that have the same publish_acl that every other user does (shown above).   It's impossible to predict which user(s) will be unlucky, and if I connect with less than a thousand or so users, no "unlucky" user gets their messages rejected. (i.e., the bug only appears under heavy load)

When I have examine at the log and identify which users are "unlucky", I can attempt to publish messages on that user's behalf via "mosquitto_pub", and the messages are rejected as unauthorized.

If I let all users disconnect (such that "vmq-admin session show" shows that no one is connected), then re-run that mosquitto_pub command, (i.e., force VerneMQ to re-authorize the user), all messages are published normally.

**Comment by DouglasWWolf** [comment]
I have more information:

When I launch the program that connects a bunch of clients to VMQ, I actually launch 8 copies of it, and each copy is responsible for connecting and managing 1000 clients.

I've discovered that this "publish not authorized" issue only happens when I have two specific copies of the program (out of the 8 possible) running.    So long as I don't run those two copies at the same time, I can run any other combination of the 8 copies and everything works fine.

In other words... this only happens if I have client-id's "sim13000 thru sim13999" and "sim14000 thru sim14999" logged in at the same time.

That smells an awful lot like a hash collision when VMQ is trying to look up the valid publish_acl for a client-id.

**Comment by ioolkos** [comment]
@DouglasWWolf thanks for opening the issue!
Interesting puzzle... :)
The question here is what do we actually hash. Is it the ACL pattern? is a cache involved for the ACL patterns? We'll have to look into those. In any case, from what I see you're not doing anything wrong and those publisher should work.

**Comment by DouglasWWolf** [comment]
Thanks for the response!

When I get some time, I am going to figure out exactly which two client-id’s don’t play well together.   (I already know one of them). 

If we can replicate this behavior with just two users, it should be reasonably simple to figure out what’s going on. 

— Doug

**Comment by larshesel** [comment]
Do you see any other errors or warnings just before the failed authorization log line? If this only happens under load, perhaps something is overloaded and times out resulting in a default `not_authorized` behaviour.

**Comment by DouglasWWolf** [comment]
Lars,

There are no other warnings in the log. 

I can now reproduce this with only 2000 users logged in, and I’m working on finding the minimum set of conditions necessary to reproduce it.  

I’m no longer convinced this is load related.   

— Doug.

**Comment by DouglasWWolf** [comment]
Guys, I have reproduced this bug with exactly two users.

Same setup as above (VerneMQ 1.7.1, authorization being done via MySQL)

Create a user called "sim13782" (use that exact string for the client-ID, username, and password).

Create a user called "sim14442" (use that exact string for the client-ID, username, and password)

Make sure they have different "publish" ACLs in the auth table.

Launch VerneMQ

Connect with user "sim13782" and publish a message or two to an authorized topic.  This message will be published as expected.

Connect with user "sim14442" and attempt to publish a message to an authorized topic.   The message will be rejected as described in the original bug description.

I can hard-reproduce this behavior every time, even after a system reboot.

Hope this helps run it down, because this is screwing up some of my test users :-)

--- Doug

**Comment by larshesel** [comment]
Thanks for the effort - we'll look into it!

**Comment by larshesel** [comment]
I'm not able to reproduce the problem. I did the following, perhaps
you can spot the difference to your approach.

First I created two users, with different publish ACLs, giving both
the right to subscribe to any topic:

```sql
INSERT INTO vmq_auth_acl
    (mountpoint, client_id, username,
     password, publish_acl, subscribe_acl)
VALUES
    ('', 'sim13782', 'sim13782', PASSWORD('sim13782'),
     '[{"pattern":"sim13782"}]',
     '[{"pattern":"#"}]');

INSERT INTO vmq_auth_acl
    (mountpoint, client_id, username,
     password, publish_acl, subscribe_acl)
VALUES
    ('', 'sim14442', 'sim14442', PASSWORD('sim14442'),
     '[{"pattern":"sim14442"}]',
     '[{"pattern":"#"}]');
```


Then I subscribe with user `sim13782` to the wildcard topic:

```shell
$ mosquitto_sub -p 1883 -i sim13782 -u sim13782 -P sim13782 -t '#' -d
Client sim13782 sending CONNECT
Client sim13782 received CONNACK (0)
Client sim13782 sending SUBSCRIBE (Mid: 1, Topic: #, QoS: 0)
Client sim13782 received SUBACK
Subscribed (mid: 1): 0
```

And then publish with user `sim14442` to the `sim14442` topic:

```shell
$ mosquitto_pub -p 1883 -i sim14442 -u sim14442 -P sim14442 -t sim14442 -m hello -d
Client sim14442 sending CONNECT
Client sim14442 received CONNACK (0)
Client sim14442 sending PUBLISH (d0, q0, r0, m1, 'sim14442', ... (5 bytes))
Client sim14442 sending DISCONNECT
```

And then in the shell where I subscribed I receive:

```shell
Client sim13782 received PUBLISH (d0, q0, r0, m0, 'sim14442', ... (5 bytes))
hello
```

It also works if I reverse the direction and `sim13782` is the
publisher and `sim14442` is the subscriber.

Can you show us the output of `select * from vmq_auth_acl where
client_id = "sim14442" OR client_id = "sim13782"; ` and also how you
publish to the topics with user `sim14442` etc?

**Comment by DouglasWWolf** [comment]
Lars,
Sorry for the delay, I was out of town.

Here is the subscribe command I run for user sim13782:

mosquitto_sub -h vmq1.harvest2o-db.com -p 8883 -i sim13782 -u sim13782 -P sim13782  --cafile \certs\amazon.crt  -t # -F "%%t %%p"


("amazon.crt" is the amazon root certificate, I'd be glad to provide a copy if it helps.)

Here is the publish command I use for user 14442:

mosquitto_pub -d -h vmq1.harvest2o-db.com -p 8883 -i sim14442 -u sim14442 -P sim14442  --cafile \certs\amazon.crt  -t node_ts/2/sim14442 -m Hello


If user sim13782 is *not* logged in to VMQ, I can execute that publish command and the publish happens as expected.


If I run the subscribe command for user sim13782 and then run that exact same publish command for user sim14442, the log shows:


2019-05-11 18:48:51.037 [error] <0.1538.0>@vmq_mqtt_fsm:auth_on_publish:718 can't auth publish [<<"sim14442">>,{[],<<"sim14442">>},0,[<<"node_ts">>,<<"2">>,<<"sim14442">>],<<"Hello">>,false] due to not_authorized

For both users (and for every other user in the system), the publish_acl field of the database is set to [{'pattern':'node_ts/+/%c'}]
Screenshot attached of the two relevant lines in the auth-table.
Thank you!
--- Doug








    On Wednesday, May 8, 2019, 11:24:57 PM PDT, Lars Hesel Christensen <notifications@github.com> wrote:  
 
 
I'm not able to reproduce the problem. I did the following, perhaps
you can spot the difference to your approach.

First I created two users, with different publish ACLs, giving both
the right to subscribe to any topic:
INSERT INTO vmq_auth_acl
    (mountpoint, client_id, username,
     password, publish_acl, subscribe_acl)
VALUES
    ('', 'sim13782', 'sim13782', PASSWORD('sim13782'),
     '[{"pattern":"sim13782"}]',
     '[{"pattern":"#"}]');

INSERT INTO vmq_auth_acl
    (mountpoint, client_id, username,
     password, publish_acl, subscribe_acl)
VALUES
    ('', 'sim14442', 'sim14442', PASSWORD('sim14442'),
     '[{"pattern":"sim14442"}]',
     '[{"pattern":"#"}]');
Then I subscribe with user sim13782 to the wildcard topic:
$ mosquitto_sub -p 1883 -i sim13782 -u sim13782 -P sim13782 -t '#' -d
Client sim13782 sending CONNECT
Client sim13782 received CONNACK (0)
Client sim13782 sending SUBSCRIBE (Mid: 1, Topic: #, QoS: 0)
Client sim13782 received SUBACK
Subscribed (mid: 1): 0
And then publish with user sim14442 to the sim14442 topic:
$ mosquitto_pub -p 1883 -i sim14442 -u sim14442 -P sim14442 -t sim14442 -m hello -d
Client sim14442 sending CONNECT
Client sim14442 received CONNACK (0)
Client sim14442 sending PUBLISH (d0, q0, r0, m1, 'sim14442', ... (5 bytes))
Client sim14442 sending DISCONNECT
And then in the shell where I subscribed I receive:
Client sim13782 received PUBLISH (d0, q0, r0, m0, 'sim14442', ... (5 bytes))
hello
It also works if I reverse the direction and sim13782 is the
publisher and sim14442 is the subscriber.

Can you show us the output of select * from vmq_auth_acl where client_id = "sim14442" OR client_id = "sim13782";  and also how you
publish to the topics with user sim14442 etc?

**Comment by DouglasWWolf** [comment]
More information, just to confirm that it's not something bizarre going on with the database:

I set up an entirely new database server from scratch.  (MySQL 8.0.11), created the VMQ auth table on that server, and populated it with the necessary users.

I re-ran the experiment with the two conflicting users ("sim13782" and "sim14442"), and the broken behavior occurred just the same as it did when using the previous database server.

To test:

1) With no other users connected, publish from sim14442... everything works as expected.

2) Have user sim13782 log in and perform a subscribe

3) Re-run the "publish" from step 1 and it fails

4) Disconnect user sim13782, re-run the publish from step 1, and it works normally.

The text of the publish and subscribe commands are verbatim as described in my previous post.

I can run all four steps above while substituting any other user for "sim13782" and everything works normally.    It is only those two specific users that don't work and play well together... which smells like a hash collision somewhere.

Hope this helps.

--- Doug

**Comment by larshesel** [comment]
Hi @DouglasWWolf I'll take a look at this now - for some reason the screenshot was not attached. Perhaps github doesn't allow that when replying via email. Can you perhaps paste it as text instead?

Thanks

**Comment by larshesel** [comment]
I'm stil not able to reproduce it so there must still be some kind of difference between our setups. Can you give us the exact sql insert commands you use to reproduce this?

**Comment by larshesel** [comment]
Hi @DouglasWWolf, I was indeed able to reproduce this after contemplating the code a bit and it indeed *is* an issue with a hash function. We're working on a fix!

**Comment by DouglasWWolf** [comment]
Lars,

Thanks much for all the effort and patience!

Any idea of when the new release with the fix will  be made available to the general public?

— D

**Comment by larshesel** [comment]
we plan to release 1.8.0 tomorrow where this patch will be part of it.

**Comment by DouglasWWolf** [comment]
Lars,

I’m incredibly impressed with depth of your knowledge here and with the assistance you’ve rendered in solving what could have been an ugly bug that affects my product’s customers.  

Thank you. 

— Doug

**Comment by larshesel** [comment]
Hi @DouglasWWolf, thank you for the kind words and thank you for your persistence and help in tracking this down. Very much appreciated!

**Comment by larshesel** [comment]
Hi @DouglasWWolf, VerneMQ 1.8.0 was just released which includes the bugfix. Let us know if that solved the issue.

**Comment by DouglasWWolf** [comment]
Thank you!
If I do a "yum install" of this new release, does it just over-write the old release, leaving my .conf file in place?
--- D

    On Wednesday, May 15, 2019, 5:42:47 AM PDT, Lars Hesel Christensen <notifications@github.com> wrote:  
 
 
Hi @DouglasWWolf, VerneMQ 1.8.0 was just released which includes the bugfix. Let us know if that solved the issue.

**Comment by larshesel** [comment]
I believe it will ask you if you want to use your config file, the new one or merge them. You probably need to merge them as there are new configs available. So best bet is likely to first make a copy of the config file, install, merge (perhaps manually, but that depends). Then you have a backup if the merge fails for some reason. (Note, I don't use yum normally, so YMMV)
